File: ValidParentheses.py
#Time Complexity: O(n)
#Space Complexity:O(n)
class Solution:
    def isValid(self, s: str) -> bool:
        stack=[]
        symbol_map={"(":")","[":"]","{":"}"}
        for i in s:
            if i in ["(","[","{"]:
                stack.append(i)
            else:
                if stack:
                    pop_val=symbol_map[stack.pop()]
                    if pop_val != i:
                        return False
                else:
                    return False
        if len(stack)>0:
            return False
        else:
            return True

File: ValidParentheses.cpp
// Time Complexity -> O(n)
// Space Complexity -> O(n) for the stack.
// Problems Faced - No!
// It runs on Leetcode!
class Solution {
public:
    bool isValid(string s) {
        int l = s.size();
        if(l%2 != 0)
            return false;
        stack<char> st;
        
        for(int i = 0; i < l; i++){
            char curr = s[i];
            if(curr == '(')
                st.push(')');
            else if(curr == '{')
                st.push('}');
            else if(curr == '[')
                st.push(']');
            else{
                if(!st.empty()){
                    char c = st.top(); st.pop();
                    if(c != curr)
                        return false;
                }else
                    return false;
            }
            }
        
        if(!st.empty())
            return false;
        return true;
    }
};